contracts/TokenHolder.sol

```diff
@@ -43,465 +42,507 @@ contract TokenHolder is MultiSigWallet {
+    {
+        ephemeralKeys[_ephemeralKey].status = AuthorizationStatus.REVOKED;
     }
 
     /**
-     * @notice public method to check if valid ephemeral key.
-     *
-     * @dev 0 spendingLimit 0 is not allowed. So spending limit greater than 0 means
-     *      ephemeralKey is present.
+     * @notice Checks if the specified key is authorized and non-expired.
      *
-     * @param _ephemeralKey Ephemeral Key which need to be checked in ephemeralKeys mapping.
+     * @param _ephemeralKey Key to check.
      *
-     * @return status is true/false.
+     * @return True if the key is currently authorized and has not expired,
+     *         otherwise false.
      */
-    function isAuthorizedEphemeralKey(
-        address _ephemeralKey
-    )
-        view
+    function isEphemeralKeyActive(address _ephemeralKey)
         public
-        returns (bool /** success status */)
+        view
+        returns (bool)
     {
-        return ephemeralKeys[_ephemeralKey].spendingLimit > 0;
+        EphemeralKeyData storage keyData = ephemeralKeys[_ephemeralKey];
+        return keyData.status == AuthorizationStatus.AUTHORIZED &&
+            keyData.expirationHeight > block.number;
     }
 
 
     /* Private Functions */
 
-    /**
-     * @notice Validate executable message signature.
-     *
-     * @param _from it will always be the contract executing the code.
-     *        It needs to be tokenholder contract address.
-     *        _from enforces TH address to be stored in wallet by sdk.
-     * @param _to the target contract the transaction will be executed upon.
-     *        e.g. Airdrop in case for Airdrop.pay.
-     * @param _nonce incremental nonce.
-     * @param _data the bytecode to be executed.
-     *         Use web3 getData method to construct _data.
-     * @param _v It's the recovery id.
-     * @param _r It's the output of an ECDSA signature.
-     * @param _s It's also the output of an ECDSA signature.
-     *
-     * @return result is true/false.
-     */
-    function verifySignature(
-        address _from,
+    function processExecutableTransaction (
+        bytes4 _callPrefix,
         address _to,
-        uint256 _nonce,
         bytes _data,
+        uint256 _nonce,
         uint8 _v,
         bytes32 _r,
         bytes32 _s
     )
         private
-        returns (EphemeralKeyData storage ephemeralKeyData_)
+        returns (bytes32 messageHash_, address key_)
     {
-        require(
-            _from == address(this),
-            "From should be tokenholder contract address"
-        );
-        require(
-            _to != address(0),
-            "to address can't be 0"
-        );
-        require(
-            _nonce != 0,
-            "Nonce can't be 0."
-        );
-        require(
-            _data.length != 0,
-            "Data length can't be 0."
+        messageHash_ = getMessageHash (
+            _callPrefix,
+            _to,
+            keccak256(_data),
+            _nonce
         );
 
```

Comment:
OpenZepplin has put few nice validations for v value. 
https://github.com/protofire/zeppelin-solidity/blob/master/contracts/ECRecovery.sol
if (v != 27 && v != 28) {
      return (address(0));
}